### PR TITLE
Service large withdrawals

### LINKED ISSUE #18: feat: modify validatorsToWithdrawFrom() to service large withdrawals

### PR DESCRIPTION
**Pull request by billythedummy**
Closes #18 and #55 (`calcWithdrawals()`)

This also applies changes relevant to other issues:
- #53 the withdrawal calculation is now all in `Numberu64`
- #62 `WithdrawalReceipt` type returned by `validatorsToWithdrawFrom()`  

Ideally we can setup localnet validators to test

### CODE REVIEW

File: src/stake-pool/instructions.ts [hunk 1]
```diff
@@ -118,7 +118,7 @@ export const depositSolInstruction = (
  * @param managerFeeAccount: Pubkey of the pool token account receiving the stake pool's fees.
  * @param poolMint: Pubkey of the pool token mint
  * @param tokenProgramId: Pubkey of the SPL token program
- * @param amount: number of pool tokens to withdraw
+ * @param dropletsToUnstake: number of droplets to withdraw
```
Comment (f8122dac91) [inline src/stake-pool/instructions.ts:121]: very nice

**Review by Johnnycus — APPROVED** [review]

**Comment by f8122dac91** [comment]
Nice! The client facing APIs are coming along nicely I think. Changes are pretty significant, will take some time to fully review this tho.

Also it looks like the two test cases have failed with `WithdrawalUnserviceableError` thrown.

File: src/stake-pool/utils.ts [hunk 2]
```diff
@@ -454,15 +515,151 @@ export async function getDefaultDepositAuthority(
...
+ * will no longer match and the result of this function will be incorrect
+ * @param lamportsToStake SOL value of the stake accounts to be staked, in lamports
+ * @param stakePool the stake pool to stake to
+ * @returns the amount of droplets (1 / 10 ** 9 scnSOL) to be received by the user
+ */
+export function calcDropletsReceivedForStakeDeposit(lamportsToStake: Numberu64, stakePool: schema.StakePool): Numberu64 {
+  return calcDropletsReceivedForDeposit(lamportsToStake, stakePool, stakePool.stakeDepositFee);
+}
+
+export type WithdrawalReceipt = {
```
Comment (f8122dac91) [inline src/stake-pool/utils.ts:559]: maybe a doc comment heading for typedoc?

File: src/stake-pool/utils.ts [hunk 3]
```diff
@@ -454,15 +515,151 @@ export async function getDefaultDepositAuthority(
...
+   * with withdrawal fees deducted
+   */
+  lamportsReceived: Numberu64;
+  /**
+   * Number of droplets paid by the user in withdrawal fees
+   */
+  dropletsFeePaid: Numberu64;
+}
+
+export type ValidatorWithdrawalReceipt = {
```
Comment (f8122dac91) [inline src/stake-pool/utils.ts:575]: maybe a doc comment heading for typedoc?

**Review by f8122dac91 — CHANGES_REQUESTED** [review]
Issues are just adding doc comments for two exported types

File: src/socean.ts [hunk 4]
```diff
@@ -17,18 +17,15 @@ import {
...
   getValidatorTransientStakeAccount,
-  calcDropletsReceivedForDeposit,
 } from "./stake-pool/utils";
 import { AccountDoesNotExistError, WalletPublicKeyUnavailableError } from "./err";
 import { cleanupRemovedValidatorsInstruction, depositSolInstruction, updateStakePoolBalanceInstruction, updateValidatorListBalanceTransaction } from "./stake-pool/instructions";
 import { WalletAdapter, TransactionSequence, TransactionSequenceSignatures, TransactionWithSigners, TRANSACTION_SEQUENCE_DEFAULT_CONFIRM_OPTIONS } from "./transactions";
 import { signAndSendTransactionSequence } from ".";
-import { StakePool } from "./stake-pool/schema";
 
 export class Socean {
```
[the remark this answers is not in the document]
Comment (billythedummy) [reply src/socean.ts:30]: nah i think its fine to stick to testnet for now, devnet stake pool needs a few more epochs to be ready

**Review by f8122dac91 — APPROVED** [review]

**Comment by f8122dac91** [comment]
All green!

**Comment by billythedummy** [comment]
```
 Transaction simulation failed: Error processing Instruction 0: Program failed to complete 
    Program 5ocnV1qiCgaQR8Jb8xWnVbApfaygJ8tNoZfgPwsgx9kx invoke [1]
    Program log: Instruction: UpdateValidatorListBalance
    Program 5ocnV1qiCgaQR8Jb8xWnVbApfaygJ8tNoZfgPwsgx9kx consumed 20326 of 200000 compute units
    Program failed to complete: Access violation in stack frame 1 at address 0x200001ee0 of size 8 by instruction #12425
    Program 5ocnV1qiCgaQR8Jb8xWnVbApfaygJ8tNoZfgPwsgx9kx failed: Program failed to complete
```

well this is new. Wait I think I've seen this error somewhere before

**Comment by Johnnycus** [comment]
> ```
>  Transaction simulation failed: Error processing Instruction 0: Program failed to complete 
>     Program 5ocnV1qiCgaQR8Jb8xWnVbApfaygJ8tNoZfgPwsgx9kx invoke [1]
>     Program log: Instruction: UpdateValidatorListBalance
>     Program 5ocnV1qiCgaQR8Jb8xWnVbApfaygJ8tNoZfgPwsgx9kx consumed 20326 of 200000 compute units
>     Program failed to complete: Access violation in stack frame 1 at address 0x200001ee0 of size 8 by instruction #12425
>     Program 5ocnV1qiCgaQR8Jb8xWnVbApfaygJ8tNoZfgPwsgx9kx failed: Program failed to complete
> ```
> 
> well this is new. Wait I think I've seen this error somewhere before

In `yarn test`? I had this a couple of times, once I had 3 failed tests, then 2, then all passed

**Comment by billythedummy** [comment]
Yeah this is quite fucked up but apparently our build is incompatible with solana build tools 1.8+. I've redeployed to devnet using a 1.7.14 docker build env and it works now: https://github.com/igneous-labs/solana-program-library/commit/2aa00d2be0a018bec776090ace9c74a10086c65e.

Oh wait i realized there is a situation where the user can start withdrawing from the reserve - there is only 1 stake account (transient or main) present for all validators in the pool, lemme modify `calcWithdrawal()` and test it

**Comment by Johnnycus** [comment]
> Yeah this is quite fucked up but apparently our build is incompatible with solana build tools 1.8+. I've redeployed to devnet using a 1.7.14 docker build env and it works now: [igneous-labs/solana-program-library@2aa00d2](https://github.com/igneous-labs/solana-program-library/commit/2aa00d2be0a018bec776090ace9c74a10086c65e).
> 
> ~Oh wait i realized there is a situation where the user can start withdrawing from the reserve - there is only 1 stake account (transient or main) present for all validators in the pool, lemme modify `calcWithdrawal()` and test it~ nvm, you need absolutely all stake accounts to be deleted

Downgrading fixes the issue?

**Comment by billythedummy** [comment]
> > Yeah this is quite fucked up but apparently our build is incompatible with solana build tools 1.8+. I've redeployed to devnet using a 1.7.14 docker build env and it works now: [igneous-labs/solana-program-library@2aa00d2](https://github.com/igneous-labs/solana-program-library/commit/2aa00d2be0a018bec776090ace9c74a10086c65e).
> > ~Oh wait i realized there is a situation where the user can start withdrawing from the reserve - there is only 1 stake account (transient or main) present for all validators in the pool, lemme modify `calcWithdrawal()` and test it~ nvm, you need absolutely all stake accounts to be deleted
> 
> Downgrading fixes the issue?

yeah downgrading the solana on-chain program build tools. Testnet wasn't affected because it was deployed a long time ago using the old build tools

**Pull request merged**